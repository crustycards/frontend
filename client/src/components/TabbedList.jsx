import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { Button } from '@material-ui/core';
import { GridList, GridTile } from 'material-ui/GridList';

class TabbedList extends Component {
  constructor (props) {
    super(props);
    this.itemsPerTab = props.itemsPerTab || 20;
    this.columns = props.columns || 4;
    this.nextTab = this.nextTab.bind(this);
    this.previousTab = this.previousTab.bind(this);
    this.state = {
      tab: 0
    };
  }

  nextTab () {
    this.setState({tab: this.state.tab + 1});
  }

  previousTab () {
    this.setState({tab: this.state.tab - 1});
  }

  render () {
    const visibleElements = [];

    let tabStart = this.state.tab * this.itemsPerTab;
    let tabEnd = tabStart + this.itemsPerTab;
    for (let i = tabStart; i < tabEnd; i++) {
      visibleElements.push(<GridTile key={i} style={{height: 'auto'}}>{this.props.elements[i]}</GridTile>);
    }

    return (
      <div className='panel'>
        <div className='center'>
          {
            this.props.elements.length > this.itemsPerTab &&
            <div>
              <Button onClick={this.previousTab} disabled={this.state.tab === 0}>Previous</Button>
              <Button onClick={this.nextTab} disabled={tabEnd >= this.props.elements.length}>Next</Button>
              <div>Tab {this.state.tab + 1} of {Math.ceil(this.props.elements.length / this.itemsPerTab)}</div>
            </div>
          }
        </div>
        <GridList children={visibleElements} cols={this.columns} cellHeight='auto' />
      </div>
    );
  }
}

TabbedList.propTypes = {
  itemsPerTab: PropTypes.number,
  columns: PropTypes.number,
  elements: PropTypes.arrayOf(PropTypes.element)
};

export default TabbedList;