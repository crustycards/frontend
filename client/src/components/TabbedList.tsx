import * as React from 'react';
import {Component} from 'react';
import {Button, GridList, GridListTile, withStyles} from '@material-ui/core';

const StyledGridListTile = withStyles({tile: {height: 'auto'}})((props: any) => (<GridListTile {...props}/>));

interface TabbedListProps {
  itemsPerTab?: number
  columns?: number
  children: JSX.Element[]
}

interface TabbedListState {
  tab: number
  itemsPerTab: number
  columns: number
}

class TabbedList extends Component<TabbedListProps, TabbedListState> {
  constructor(props: TabbedListProps) {
    super(props);
    this.nextTab = this.nextTab.bind(this);
    this.previousTab = this.previousTab.bind(this);
    this.state = {
      tab: 0,
      itemsPerTab: props.itemsPerTab || 20,
      columns: props.columns || 4
    };
  }

  nextTab() {
    if (this.canGoToNextPage()) {
      this.setState({tab: this.state.tab + 1});
    }
  }

  previousTab() {
    if (this.canGoToPreviousPage()) {
      this.setState({tab: this.state.tab - 1});
    }
  }

  getTabRange() {
    const tabStart = this.state.tab * this.state.itemsPerTab;
    const tabEnd = tabStart + this.state.itemsPerTab;
    return {
      tabStart,
      tabEnd
    };
  }

  canGoToNextPage() {
    const {tabEnd} = this.getTabRange();
    return tabEnd < this.props.children.length;
  }

  canGoToPreviousPage() {
    return this.state.tab > 0;
  }

  render() {
    const visibleElements = [];

    const {tabStart, tabEnd} = this.getTabRange();
    // TODO - Convert for loop to map
    for (let i = tabStart; i < tabEnd; i++) {
      if (!this.props.children[i]) {
        break;
      }
      visibleElements.push(
        <StyledGridListTile
          key={i}
        >
          {this.props.children[i]}
        </StyledGridListTile>
      );
    }

    return (
      <div className='panel'>
        <div className='center'>
          {
            this.props.children.length > this.state.itemsPerTab &&
            <div>
              <Button
                className={'previous-tab'}
                onClick={this.previousTab}
                disabled={!this.canGoToPreviousPage()}
              >
                Previous
              </Button>
              <Button
                className={'next-tab'}
                onClick={this.nextTab}
                disabled={!this.canGoToNextPage()}
              >
                Next
              </Button>
              <div>
                Tab {
                  this.state.tab + 1
                } of {
                  Math.ceil(this.props.children.length / this.state.itemsPerTab)
                }
              </div>
            </div>
          }
        </div>
        <GridList cols={this.state.columns} cellHeight={'auto'}>
          {visibleElements}
        </GridList>
      </div>
    );
  }
}

export default TabbedList;
