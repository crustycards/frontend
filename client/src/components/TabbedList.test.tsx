import * as React from 'react';
import TabbedList from './TabbedList';
import {shallow, mount} from 'enzyme';
import {GridListTile} from '@material-ui/core';

it('matches snapshot of single page of elements', () => {
  const wrapper = shallow(<TabbedList>{[
    <div>Element One</div>,
    <div>Element Two</div>,
    <div>Element Three</div>,
    <div>Element Four</div>
  ]}</TabbedList>);

  expect(wrapper).toMatchSnapshot();
});

it('renders correct number of elements', () => {
  const wrapper = mount(<TabbedList>{[
    <div>Element One</div>,
    <div>Element Two</div>,
    <div>Element Three</div>,
    <div>Element Four</div>
  ]}</TabbedList>);

  expect(wrapper.find(GridListTile).length).toEqual(4);
});

it('renders no more than 20 elements at once', () => {
  const wrapper = mount(<TabbedList>{[
    <div>Element One</div>,
    <div>Element Two</div>,
    <div>Element Three</div>,
    <div>Element Four</div>,
    <div>Element Five</div>,
    <div>Element Six</div>,
    <div>Element Seven</div>,
    <div>Element Eight</div>,
    <div>Element Nine</div>,
    <div>Element Ten</div>,
    <div>Element Eleven</div>,
    <div>Element Twelve</div>,
    <div>Element Thirteen</div>,
    <div>Element Fourteen</div>,
    <div>Element Fifteen</div>,
    <div>Element Sixteen</div>,
    <div>Element Seventeen</div>,
    <div>Element Eighteen</div>,
    <div>Element Nineteen</div>,
    <div>Element Twenty</div>,
    <div>Element Twenty One</div>,
    <div>Element Twenty Two</div>
  ]}</TabbedList>);

  expect(wrapper.find(GridListTile).length).toEqual(20);
});