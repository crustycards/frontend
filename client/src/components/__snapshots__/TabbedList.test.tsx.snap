// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot of single page of elements 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <TabbedList>
    <div>
      Element One
    </div>
    <div>
      Element Two
    </div>
    <div>
      Element Three
    </div>
    <div>
      Element Four
    </div>
  </TabbedList>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <div
          className="center"
        />,
        <WithStyles(GridList)
          cellHeight="auto"
          cols={4}
        >
          <WithStyles(Component)>
            <div>
              Element One
            </div>
          </WithStyles(Component)>
          <WithStyles(Component)>
            <div>
              Element Two
            </div>
          </WithStyles(Component)>
          <WithStyles(Component)>
            <div>
              Element Three
            </div>
          </WithStyles(Component)>
          <WithStyles(Component)>
            <div>
              Element Four
            </div>
          </WithStyles(Component)>
        </WithStyles(GridList)>,
      ],
      "className": "panel",
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": false,
          "className": "center",
        },
        "ref": null,
        "rendered": false,
        "type": "div",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "cellHeight": "auto",
          "children": Array [
            <WithStyles(Component)>
              <div>
                Element One
              </div>
            </WithStyles(Component)>,
            <WithStyles(Component)>
              <div>
                Element Two
              </div>
            </WithStyles(Component)>,
            <WithStyles(Component)>
              <div>
                Element Three
              </div>
            </WithStyles(Component)>,
            <WithStyles(Component)>
              <div>
                Element Four
              </div>
            </WithStyles(Component)>,
          ],
          "cols": 4,
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": "0",
            "nodeType": "class",
            "props": Object {
              "children": <div>
                Element One
              </div>,
            },
            "ref": null,
            "rendered": Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": "Element One",
              },
              "ref": null,
              "rendered": "Element One",
              "type": "div",
            },
            "type": [Function],
          },
          Object {
            "instance": null,
            "key": "1",
            "nodeType": "class",
            "props": Object {
              "children": <div>
                Element Two
              </div>,
            },
            "ref": null,
            "rendered": Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": "Element Two",
              },
              "ref": null,
              "rendered": "Element Two",
              "type": "div",
            },
            "type": [Function],
          },
          Object {
            "instance": null,
            "key": "2",
            "nodeType": "class",
            "props": Object {
              "children": <div>
                Element Three
              </div>,
            },
            "ref": null,
            "rendered": Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": "Element Three",
              },
              "ref": null,
              "rendered": "Element Three",
              "type": "div",
            },
            "type": [Function],
          },
          Object {
            "instance": null,
            "key": "3",
            "nodeType": "class",
            "props": Object {
              "children": <div>
                Element Four
              </div>,
            },
            "ref": null,
            "rendered": Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": "Element Four",
              },
              "ref": null,
              "rendered": "Element Four",
              "type": "div",
            },
            "type": [Function],
          },
        ],
        "type": [Function],
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <div
            className="center"
          />,
          <WithStyles(GridList)
            cellHeight="auto"
            cols={4}
          >
            <WithStyles(Component)>
              <div>
                Element One
              </div>
            </WithStyles(Component)>
            <WithStyles(Component)>
              <div>
                Element Two
              </div>
            </WithStyles(Component)>
            <WithStyles(Component)>
              <div>
                Element Three
              </div>
            </WithStyles(Component)>
            <WithStyles(Component)>
              <div>
                Element Four
              </div>
            </WithStyles(Component)>
          </WithStyles(GridList)>,
        ],
        "className": "panel",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": false,
            "className": "center",
          },
          "ref": null,
          "rendered": false,
          "type": "div",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "cellHeight": "auto",
            "children": Array [
              <WithStyles(Component)>
                <div>
                  Element One
                </div>
              </WithStyles(Component)>,
              <WithStyles(Component)>
                <div>
                  Element Two
                </div>
              </WithStyles(Component)>,
              <WithStyles(Component)>
                <div>
                  Element Three
                </div>
              </WithStyles(Component)>,
              <WithStyles(Component)>
                <div>
                  Element Four
                </div>
              </WithStyles(Component)>,
            ],
            "cols": 4,
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": "0",
              "nodeType": "class",
              "props": Object {
                "children": <div>
                  Element One
                </div>,
              },
              "ref": null,
              "rendered": Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {
                  "children": "Element One",
                },
                "ref": null,
                "rendered": "Element One",
                "type": "div",
              },
              "type": [Function],
            },
            Object {
              "instance": null,
              "key": "1",
              "nodeType": "class",
              "props": Object {
                "children": <div>
                  Element Two
                </div>,
              },
              "ref": null,
              "rendered": Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {
                  "children": "Element Two",
                },
                "ref": null,
                "rendered": "Element Two",
                "type": "div",
              },
              "type": [Function],
            },
            Object {
              "instance": null,
              "key": "2",
              "nodeType": "class",
              "props": Object {
                "children": <div>
                  Element Three
                </div>,
              },
              "ref": null,
              "rendered": Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {
                  "children": "Element Three",
                },
                "ref": null,
                "rendered": "Element Three",
                "type": "div",
              },
              "type": [Function],
            },
            Object {
              "instance": null,
              "key": "3",
              "nodeType": "class",
              "props": Object {
                "children": <div>
                  Element Four
                </div>,
              },
              "ref": null,
              "rendered": Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {
                  "children": "Element Four",
                },
                "ref": null,
                "rendered": "Element Four",
                "type": "div",
              },
              "type": [Function],
            },
          ],
          "type": [Function],
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
    "attachTo": undefined,
    "hydrateIn": undefined,
  },
}
`;
